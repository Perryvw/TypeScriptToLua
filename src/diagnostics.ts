import * as ts from "typescript";
import { TranspileError } from "./TranspileError";

export const transpileError = (error: TranspileError): ts.Diagnostic => ({
    file: error.node.getSourceFile(),
    start: error.node.getStart(),
    length: error.node.getWidth(),
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: error.message,
});

export const tstlOptionsAreMovingToTheTstlObject = (tstl: Record<string, any>): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Warning,
    code: 0,
    source: "typescript-to-lua",
    messageText:
        'TSTL options are moving to the "tstl" object. Adjust your tsconfig to look like\n' +
        `"tstl": ${JSON.stringify(tstl, undefined, 4)}`,
});

export const toLoadTransformerItShouldBeTranspiled = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `To load "${transform}" transformer it should be transpiled or "ts-node" should be installed`,
});

export const couldNotResolveTransformerFrom = (transform: string, base: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `Could not resolve "${transform}" transformer from "${base}".`,
});

export const transformerShouldHaveAExport = (transform: string, importName: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should have a "${importName}" export`,
});

export const transformerShouldBeATsTransformerFactory = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should be a ts.TransformerFactory or an object with ts.TransformerFactory values`,
});

export const watchErrorSummary = (errorCount: number): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Message,
    code: errorCount === 1 ? 6193 : 6194,
    messageText:
        errorCount === 1
            ? "Found 1 error. Watching for file changes."
            : `Found ${errorCount} errors. Watching for file changes.`,
});

const createCommandLineError = <Args extends any[]>(code: number, getMessage: (...args: Args) => string) => (
    ...args: Args
): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code,
    messageText: getMessage(...args),
});

export const unknownCompilerOption = createCommandLineError(
    5023,
    (name: string) => `Unknown compiler option '${name}'.`
);

export const compilerOptionRequiresAValueOfType = createCommandLineError(
    5024,
    (name: string, type: string) => `Compiler option '${name}' requires a value of type ${type}.`
);

export const optionProjectCannotBeMixedWithSourceFilesOnACommandLine = createCommandLineError(
    5042,
    () => "Option 'project' cannot be mixed with source files on a command line."
);

export const cannotFindATsconfigJsonAtTheSpecifiedDirectory = createCommandLineError(
    5057,
    (dir: string) => `Cannot find a tsconfig.json file at the specified directory: '${dir}'.`
);

export const theSpecifiedPathDoesNotExist = createCommandLineError(
    5058,
    (dir: string) => `The specified path does not exist: '${dir}'.`
);

export const compilerOptionExpectsAnArgument = createCommandLineError(
    6044,
    (name: string) => `Compiler option '${name}' expects an argument.`
);

export const argumentForOptionMustBe = createCommandLineError(
    6046,
    (name: string, values: string) => `Argument for '${name}' option must be: ${values}.`
);

export const optionCanOnlyBeSpecifiedInTsconfigJsonFile = createCommandLineError(
    6064,
    (name: string) => `Option '${name}' can only be specified in 'tsconfig.json' file.`
);

export const optionBuildMustBeFirstCommandLineArgument = createCommandLineError(
    6369,
    () => "Option '--build' must be the first command line argument."
);
