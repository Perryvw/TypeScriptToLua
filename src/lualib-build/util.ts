import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
