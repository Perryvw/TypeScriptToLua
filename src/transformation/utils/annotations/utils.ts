import * as ts from "typescript";
import { AnnotationKind, getSignatureAnnotations, getTypeAnnotations } from ".";
import { TransformationContext } from "../../context";
import { findFirstNodeAbove, inferAssignedType } from "../typescript";

export function isTupleReturnCall(context: TransformationContext, node: ts.Node): boolean {
    if (!ts.isCallExpression(node)) {
        return false;
    }

    const signature = context.checker.getResolvedSignature(node);
    if (signature) {
        if (getSignatureAnnotations(context, signature).has(AnnotationKind.TupleReturn)) {
            return true;
        }

        // Only check function type for directive if it is declared as an interface or type alias
        const declaration = signature.getDeclaration();
        const isInterfaceOrAlias =
            declaration &&
            declaration.parent &&
            ((ts.isInterfaceDeclaration(declaration.parent) && ts.isCallSignatureDeclaration(declaration)) ||
                ts.isTypeAliasDeclaration(declaration.parent));
        if (!isInterfaceOrAlias) {
            return false;
        }
    }

    const type = context.checker.getTypeAtLocation(node.expression);
    return getTypeAnnotations(context, type).has(AnnotationKind.TupleReturn);
}

export function isInTupleReturnFunction(context: TransformationContext, node: ts.Node): boolean {
    const declaration = findFirstNodeAbove(node, ts.isFunctionLike);
    if (!declaration) {
        return false;
    }

    let functionType: ts.Type | undefined;
    if (ts.isFunctionExpression(declaration) || ts.isArrowFunction(declaration)) {
        functionType = inferAssignedType(context, declaration);
    } else if (ts.isMethodDeclaration(declaration) && ts.isObjectLiteralExpression(declaration.parent)) {
        // Manually lookup type for object literal properties declared with method syntax
        const interfaceType = inferAssignedType(context, declaration.parent);
        const propertySymbol = interfaceType.getProperty(declaration.name.getText());
        if (propertySymbol) {
            functionType = context.checker.getTypeOfSymbolAtLocation(propertySymbol, declaration);
        }
    }

    if (functionType === undefined) {
        functionType = context.checker.getTypeAtLocation(declaration);
    }

    // Check all overloads for directive
    const signatures = functionType.getCallSignatures();
    if (signatures && signatures.some(s => getSignatureAnnotations(context, s).has(AnnotationKind.TupleReturn))) {
        return true;
    }

    return getTypeAnnotations(context, functionType).has(AnnotationKind.TupleReturn);
}

export function isLuaIteratorType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(context, type).has(AnnotationKind.LuaIterator);
}

export function isVarArgType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(context, type).has(AnnotationKind.VarArg);
}

export function isForRangeType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(context, type).has(AnnotationKind.ForRange);
}
