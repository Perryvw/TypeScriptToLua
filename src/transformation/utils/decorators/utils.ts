import * as ts from "typescript";
import { TransformationContext } from "../../context";
import { DecoratorKind, getCustomDecorators, getCustomSignatureDecorators } from "../decorators";
import { findFirstNodeAbove, inferAssignedType } from "../typescript";

export function isTupleReturnCall(context: TransformationContext, node: ts.Node): boolean {
    if (!ts.isCallExpression(node)) {
        return false;
    }

    const signature = context.checker.getResolvedSignature(node);
    if (signature) {
        if (getCustomSignatureDecorators(context, signature).has(DecoratorKind.TupleReturn)) {
            return true;
        }

        // Only check function type for directive if it is declared as an interface or type alias
        const declaration = signature.getDeclaration();
        const isInterfaceOrAlias =
            declaration &&
            declaration.parent &&
            ((ts.isInterfaceDeclaration(declaration.parent) && ts.isCallSignatureDeclaration(declaration)) ||
                ts.isTypeAliasDeclaration(declaration.parent));
        if (!isInterfaceOrAlias) {
            return false;
        }
    }

    const type = context.checker.getTypeAtLocation(node.expression);
    return getCustomDecorators(context, type).has(DecoratorKind.TupleReturn);
}

export function isInTupleReturnFunction(context: TransformationContext, node: ts.Node): boolean {
    const declaration = findFirstNodeAbove(node, ts.isFunctionLike);
    if (!declaration) {
        return false;
    }

    let functionType: ts.Type | undefined;
    if (ts.isFunctionExpression(declaration) || ts.isArrowFunction(declaration)) {
        functionType = inferAssignedType(context, declaration);
    } else if (ts.isMethodDeclaration(declaration) && ts.isObjectLiteralExpression(declaration.parent)) {
        // Manually lookup type for object literal properties declared with method syntax
        const interfaceType = inferAssignedType(context, declaration.parent);
        const propertySymbol = interfaceType.getProperty(declaration.name.getText());
        if (propertySymbol) {
            functionType = context.checker.getTypeOfSymbolAtLocation(propertySymbol, declaration);
        }
    }

    if (functionType === undefined) {
        functionType = context.checker.getTypeAtLocation(declaration);
    }

    // Check all overloads for directive
    const signatures = functionType.getCallSignatures();
    if (signatures && signatures.some(s => getCustomSignatureDecorators(context, s).has(DecoratorKind.TupleReturn))) {
        return true;
    }

    const decorators = getCustomDecorators(context, functionType);
    return decorators.has(DecoratorKind.TupleReturn);
}

export function isLuaIteratorType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getCustomDecorators(context, type).has(DecoratorKind.LuaIterator);
}

export function isVarArgType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getCustomDecorators(context, type).has(DecoratorKind.Vararg);
}

export function isForRangeType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getCustomDecorators(context, type).has(DecoratorKind.ForRange);
}
