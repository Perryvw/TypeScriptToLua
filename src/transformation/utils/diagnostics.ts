import * as ts from "typescript";

const createDiagnosticFactory = <TArgs extends any[]>(
    message: string | ((...args: TArgs) => string),
    category = ts.DiagnosticCategory.Error
) => (node: ts.Node, ...args: TArgs): ts.Diagnostic => ({
    file: node.getSourceFile(),
    start: node.getStart(),
    length: node.getWidth(),
    category,
    code: 0,
    source: "typescript-to-lua",
    messageText: typeof message === "string" ? message : message(...args),
});

export const forbiddenForIn = createDiagnosticFactory(`Iterating over arrays with 'for ... in' is not allowed.`);

export const invalidForRangeCall = createDiagnosticFactory((message: string) => `Invalid @forRange call: ${message}.`);
