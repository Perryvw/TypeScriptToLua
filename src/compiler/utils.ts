import { create as createResolve } from "enhanced-resolve";
import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import { assert } from "../utils";
import * as diagnosticFactories from "./diagnostics";

// https://github.com/webpack/enhanced-resolve/blob/0001f80dacf033ac4a0e690b2766e0965c458266/lib/Resolver.js#L280-L288
export const isResolveError = (error: unknown): error is Error & { details: string } =>
    error instanceof Error && "details" in error;

const resolveConfigImport = createResolve.sync({ extensions: [".js", ".ts", ".tsx"] });

export function loadConfigImport(
    kind: string,
    optionName: string,
    basedir: string,
    query: string,
    importName = "default"
): { error?: ts.Diagnostic; result?: unknown } {
    if (typeof query !== "string") {
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "string") };
    }

    let resolved: string;
    try {
        const result = resolveConfigImport({}, basedir, query);
        assert(typeof result === "string");
        resolved = result;
    } catch (error) {
        if (!isResolveError(error)) throw error;
        return { error: diagnosticFactories.couldNotResolveFrom(kind, query, basedir) };
    }

    const hasNoRequireHook = require.extensions[".ts"] === undefined;
    if (hasNoRequireHook && (resolved.endsWith(".ts") || resolved.endsWith(".tsx"))) {
        try {
            require("ts-node/register/transpile-only");
        } catch (err) {
            if (err.code !== "MODULE_NOT_FOUND") throw err;
            return { error: diagnosticFactories.toLoadItShouldBeTranspiled(kind, query) };
        }
    }

    const commonjsModule = require(resolved);
    const factoryModule = commonjsModule.__esModule ? commonjsModule : { default: commonjsModule };
    const result = factoryModule[importName];
    if (result === undefined) {
        return { error: diagnosticFactories.shouldHaveAExport(kind, query, importName) };
    }

    return { result };
}
