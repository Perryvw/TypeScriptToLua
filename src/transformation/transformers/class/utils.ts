import * as ts from "typescript";
import { DecoratorKind, getCustomDecorators } from "../../utils/decorators";
import { TransformationContext } from "../../context";

export function isStaticNode(node: ts.Node): boolean {
    return node.modifiers !== undefined && node.modifiers.some(m => m.kind === ts.SyntaxKind.StaticKeyword);
}

export function getExtendedTypeNode(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.ExpressionWithTypeArguments | undefined {
    if (node && node.heritageClauses) {
        for (const clause of node.heritageClauses) {
            if (clause.token === ts.SyntaxKind.ExtendsKeyword) {
                const superType = context.checker.getTypeAtLocation(clause.types[0]);
                const decorators = getCustomDecorators(context, superType);
                if (!decorators.has(DecoratorKind.PureAbstract)) {
                    return clause.types[0];
                }
            }
        }
    }
}

export function getExtendedType(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.Type | undefined {
    const extendedTypeNode = getExtendedTypeNode(context, node);
    return extendedTypeNode && context.checker.getTypeAtLocation(extendedTypeNode);
}

export function findInClassOrAncestor(
    context: TransformationContext,
    classDeclaration: ts.ClassLikeDeclarationBase,
    callback: (classDeclaration: ts.ClassLikeDeclarationBase) => boolean
): ts.ClassLikeDeclarationBase | undefined {
    if (callback(classDeclaration)) {
        return classDeclaration;
    }

    const extendsType = getExtendedType(context, classDeclaration);
    if (!extendsType) {
        return undefined;
    }

    const symbol = extendsType.getSymbol();
    if (symbol === undefined) {
        return undefined;
    }

    const symbolDeclarations = symbol.getDeclarations();
    if (symbolDeclarations === undefined) {
        return undefined;
    }

    const declaration = symbolDeclarations.find(ts.isClassLike);
    if (!declaration) {
        return undefined;
    }

    return findInClassOrAncestor(context, declaration, callback);
}
