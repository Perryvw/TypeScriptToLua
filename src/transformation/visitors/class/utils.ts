import * as ts from "typescript";
import { TransformationContext } from "../../context";
import { AnnotationKind, getTypeAnnotations } from "../../utils/annotations";

export function isStaticNode(node: ts.Node): boolean {
    return (node.modifiers ?? []).some(m => m.kind === ts.SyntaxKind.StaticKeyword);
}

export function getExtendedTypeNode(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.ExpressionWithTypeArguments | undefined {
    if (node && node.heritageClauses) {
        for (const clause of node.heritageClauses) {
            if (clause.token === ts.SyntaxKind.ExtendsKeyword) {
                const superType = context.checker.getTypeAtLocation(clause.types[0]);
                const annotations = getTypeAnnotations(context, superType);
                if (!annotations.has(AnnotationKind.PureAbstract)) {
                    return clause.types[0];
                }
            }
        }
    }
}

export function getExtendedType(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.Type | undefined {
    const extendedTypeNode = getExtendedTypeNode(context, node);
    return extendedTypeNode && context.checker.getTypeAtLocation(extendedTypeNode);
}
