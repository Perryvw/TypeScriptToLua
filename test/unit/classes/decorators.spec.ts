import { decoratorInvalidContext } from "../../../src/transformation/utils/diagnostics";
import * as util from "../../util";

test("Class decorator with no parameters", () => {
    util.testFunction`
        function setBool<T extends new (...args: any[]) => {}>(constructor: T) {
            return class extends constructor {
                decoratorBool = true;
            }
        }

        @setBool
        class TestClass {
            public decoratorBool = false;
        }

        return new TestClass();
    `.expectToMatchJsResult();
});

test("Class decorator with parameters", () => {
    util.testFunction`
        function setNum(numArg: number) {
            return <T extends new (...args: any[]) => {}>(constructor: T) => {
                return class extends constructor {
                    decoratorNum = numArg;
                };
            };
        }

        @setNum(420)
        class TestClass {
            public decoratorNum;
        }

        return new TestClass();
    `.expectToMatchJsResult();
});

test("Multiple class decorators", () => {
    util.testFunction`
        function setTen<T extends new (...args: any[]) => {}>(constructor: T) {
            return class extends constructor {
                decoratorTen = 10;
            }
        }

        function setNum<T extends new (...args: any[]) => {}>(constructor: T) {
            return class extends constructor {
                decoratorNum = 410;
            }
        }

        @setTen
        @setNum
        class TestClass {
            public decoratorTen;
            public decoratorNum;
        }

        return new TestClass();
    `.expectToMatchJsResult();
});

test("Class decorator with inheritance", () => {
    util.testFunction`
        function setTen<T extends new (...args: any[]) => {}>(constructor: T) {
            return class extends constructor {
                decoratorTen = 10;
            }
        }

        class TestClass {
            public decoratorTen = 0;
        }

        @setTen
        class SubTestClass extends TestClass {
            public decoratorTen = 5;
        }

        return new SubTestClass();
    `.expectToMatchJsResult();
});

test("Class decorators are applied in order and executed in reverse order", () => {
    util.testFunction`
        const order = [];

        function pushOrder(index: number) {
            order.push("eval " + index);
            return (constructor: new (...args: any[]) => {}) => {
                order.push("execute " + index);
            };
        }

        @pushOrder(1)
        @pushOrder(2)
        @pushOrder(3)
        class TestClass {}

        return order;
    `.expectToMatchJsResult();
});

test("Throws error if decorator function has void context", () => {
    util.testFunction`
        function decorator(this: void, constructor: new (...args: any[]) => {}) {}

        @decorator
        class TestClass {}
    `.expectDiagnosticsToMatchSnapshot([decoratorInvalidContext.code]);
});

test("Exported class decorator", () => {
    util.testModule`
        function decorator<T extends new (...args: any[]) => any>(Class: T): T {
            return class extends Class {
                public bar = "foobar";
            };
        }

        @decorator
        export class Foo {}
    `
        .setReturnExport("Foo", "bar")
        .expectToMatchJsResult();
});

test.each([
    ["method() {}"],
    ["property;"],
    ["propertyWithInitializer = () => {};"],
    ["['evaluated property'];"],
    ["get getter() { return 5 }"],
    ["set setter(value) {}"],
    ["static method() {}"],
    ["static property;"],
    ["static propertyWithInitializer = () => {}"],
    ["static get getter() { return 5 }"],
    ["static set setter(value) {}"],
    ["static ['evaluated property'];"],
])("Decorate class member (%p)", memberStatement => {
    util.testFunction`
        let decoratorTarget: any | undefined;
        let decoratorTargetKey: string | undefined;

        const decorator = (target, key) => {
            decoratorTarget = target;
            decoratorTargetKey = key;
        };

        class Foo {
            @decorator
            ${memberStatement}
        }

        const targetKind = decoratorTarget === Foo.prototype ? "prototype" : "class";
        return [targetKind, decoratorTargetKey];
    `.expectToMatchJsResult();
});

test.each([["method(@decorator a) {}"], ["static method(@decorator a) {}"]])(
    "Decorate method parameter (%p)",
    methodStatement => {
        util.testFunction`
            let decoratorTarget;
            let decoratorTargetKey;
            let decoratorTargetKeyIndex;

            const decorator = (target, key, index) => {
                decoratorTarget = target;
                decoratorTargetKey = key;
                decoratorTargetKeyIndex = index;
            };

            class Foo {
                ${methodStatement}
            }

            const targetKind = decoratorTarget === Foo.prototype ? "prototype" : "class";
            return [targetKind, decoratorTargetKey, decoratorTargetKeyIndex];
        `.expectToMatchJsResult();
    }
);
