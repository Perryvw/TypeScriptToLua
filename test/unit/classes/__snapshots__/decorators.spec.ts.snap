// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Throws error if decorator function has void context: code 1`] = `
"local ____lualib = require(\\"lualib_bundle\\")
local __TS__Class = ____lualib.__TS__Class
local __TS__Decorate = ____lualib.__TS__Decorate
local ____exports = {}
function ____exports.__main(self)
    local function decorator(constructor)
    end
    local TestClass = __TS__Class()
    TestClass.name = \\"TestClass\\"
    function TestClass.prototype.____constructor(self)
    end
    TestClass = __TS__Decorate({decorator}, TestClass)
end
return ____exports"
`;

exports[`Throws error if decorator function has void context: diagnostics 1`] = `"main.ts(4,9): error TSTL: Decorator function cannot have 'this: void'."`;
